cli/module_graph.rs

```diff
@@ -21,20 +24,406 @@ use serde::Serialize;
+  Path,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct TsReferenceDesc {
+  pub kind: TsReferenceKind,
+  pub specifier: String,
+  pub location: Location,
+}
+
+// TODO(bartlomieju): maybe should live in `tsc.rs`?
+// TODO(bartlomieju): handle imports in ambient contexts/TS modules
+/// This function is a port of `ts.preProcessFile()`
+///
+/// Additionally it captures `@deno-types` references directly
+/// preceeding `import .. from` and `export .. from` statements.
+pub fn pre_process_file(
+  file_name: &str,
+  media_type: MediaType,
+  source_code: &str,
+  analyze_dynamic_imports: bool,
+) -> Result<(Vec<ImportDesc>, Vec<TsReferenceDesc>), SwcDiagnosticBuffer> {
+  let parser = AstParser::new();
+  parser.parse_module(file_name, media_type, source_code, |parse_result| {
+    let module = parse_result?;
+    let mut collector = DependencyVisitor {
+      dependencies: vec![],
+    };
+    let module_span = module.span;
+    collector.visit_module(&module, &module);
+
+    let dependency_descriptors = collector.dependencies;
+
+    // for each import check if there's relevant @deno-types directive
+    let imports = dependency_descriptors
+      .iter()
+      .filter(|desc| {
+        if analyze_dynamic_imports {
+          return true;
+        }
+
+        desc.kind != DependencyKind::DynamicImport
+      })
+      .map(|desc| {
+        let location = parser.get_span_location(desc.span);
+        let deno_types = get_deno_types(&parser, desc.span);
+        ImportDesc {
+          specifier: desc.specifier.to_string(),
+          deno_types,
+          location: location.into(),
+        }
+      })
+      .collect();
+
+    // analyze comment from beginning of the file and find TS directives
+    let comments = parser
+      .comments
+      .take_leading_comments(module_span.lo())
+      .unwrap_or_else(Vec::new);
+
+    let mut references = vec![];
+    for comment in comments {
+      if comment.kind != CommentKind::Line {
+        continue;
+      }
+
+      // TODO(bartlomieju): you can do better than that...
+      let text = comment.text.to_string();
+      let (kind, specifier_in_quotes) =
+        if text.starts_with("/ <reference path=") {
+          (
+            TsReferenceKind::Path,
+            text.trim_start_matches("/ <reference path="),
+          )
+        } else if text.starts_with("/ <reference lib=") {
+          (
+            TsReferenceKind::Lib,
+            text.trim_start_matches("/ <reference lib="),
+          )
+        } else if text.starts_with("/ <reference types=") {
+          (
+            TsReferenceKind::Types,
+            text.trim_start_matches("/ <reference types="),
+          )
+        } else {
+          continue;
+        };
+      let specifier = specifier_in_quotes
+        .trim_end_matches("/>")
+        .trim_end()
+        .trim_start_matches('\"')
+        .trim_start_matches('\'')
+        .trim_end_matches('\"')
+        .trim_end_matches('\'')
+        .to_string();
```

Comment:
This code parsing reference tags should be split out into a standalone function with its own unit tests.

Reply:
Done